AK/BigIntBase.h

```diff
@@ -503,39 +560,108 @@ struct StorageOperations {
         if (size2 < size) {
             if (sign1) {
                 auto inverted = buffer.allocate(size1);
-                negate(operand1, UnsignedStorageSpan { inverted, size1 });
+                negate(operand1, StorageSpan<WordType, false> { inverted, size1 });
                 data1 = inverted;
             }
             if (sign2) {
                 auto inverted = buffer.allocate(size2);
-                negate(operand2, UnsignedStorageSpan { inverted, size2 });
+                negate(operand2, StorageSpan<WordType, false> { inverted, size2 });
                 data2 = inverted;
             }
         }
         size1 = min(size1, size), size2 = min(size2, size);
 
         // Do schoolbook O(size1 * size2).
-        DoubleWord carry = 0;
+        DoubleWordType carry = 0;
         for (size_t i = 0; i < size; ++i) {
-            result[i] = static_cast<NativeWord>(carry);
+            result[i] = static_cast<WordType>(carry);
             carry >>= word_size;
 
             size_t start_index = i >= size2 ? i - size2 + 1 : 0;
             size_t end_index = min(i + 1, size1);
 
             for (size_t j = start_index; j < end_index; ++j) {
-                auto x = static_cast<DoubleWord>(data1[j]) * data2[i - j];
+                auto x = static_cast<DoubleWordType>(data1[j]) * data2[i - j];
 
                 bool ncarry = false;
-                result[i] = add_words(result[i], static_cast<NativeWord>(x), ncarry);
+                result[i] = add_words(result[i], static_cast<WordType>(x), ncarry);
                 carry += (x >> word_size) + ncarry;
             }
         }
 
         if (size2 < size && (sign1 ^ sign2))
             negate(result, result);
     }
+
+    template<bool restore_remainder = false>
+    static constexpr void div_mod_internal(
+        StorageSpan<WordType, false> dividend, StorageSpan<WordType, false> divisor,
```

Comment:
Other functions here take `IntegerReadonlyStorage<WordType> auto` instead of `StorageSpan`s. The idea was to force the compiler to unroll loops and inline functions when dealing with small fixed big ints.

Reply:
I have thought about this a bit
So Span or StargeSpan for that matter is fully constexpr, so I think this should not hinder unrolling, when the span size is a compile time constant,
I havent checked this with compiler explorer though
additionally, allowing plain StaticStorage instead would't allow me to use the nice and explicit `trim` and `slice` helpers, but would revert the code to less legible pointer arithmatic

Reply:
Oh wow, division generates so much code that the compiler even decides to not bother inlining shifts into div_mod_internal: https://godbolt.org/z/dMqP4K6dM. So then I think it is safe to assume that replacing spans with templated types won't matter much.